core/kernel/msg_param.c

```diff
@@ -125,6 +125,12 @@ struct mobj *msg_param_mobj_from_noncontig(paddr_t buf_ptr, size_t size,
 	page_offset = buf_ptr & SMALL_PAGE_MASK;
 	if (ADD_OVERFLOW(size, page_offset, &size_plus_offs))
 		return NULL;
+
+	if (!size_plus_offs) {
```

Comment:
I'll switch that to a check on size and move it above then

Reply:
The issue is that if we reject zero sized buffer, a lot of existing and deployed U-Boot will fail to probe their optee driver. A fix was sent to U-Boot to prevent that, see [[PATCH 2/2] tee: optee: don't enumerate services if there ain't any](https://lore.kernel.org/u-boot/20231129123753.1906956-2-etienne.carriere@foss.st.com/#r).

I would be more in favor to relax OP-TEE and allow zero sized buffer to be registered, as before, for compat with existing U-Boot implementation:
```patch
 	size_t size_plus_offs = 0;
 	size_t msize = 0;
 
+	if (!size) {
+		/* Extent to 1 byte to allow registering of the referred page */
+		size = 1;
+	}
+
 	page_offset = buf_ptr & SMALL_PAGE_MASK;
 	if (ADD_OVERFLOW(size, page_offset, &size_plus_offs))
 		return NULL;
```

Reply:
U-Boot will boot if and only if the page_offset is !=0 when the size is 0, it will crash otherwise. Today's U-Boot success to boot is random if enumerating services returns a 0-sized list. (Only dependent of the page offset)

Looking at your patches, it seems like the kind of fix to apply. Is U-Boot compatibility enough to not implement te right fix in OP-TEE

Reply:
i think U-Boot should be fixed (patches sent to U-Boot mailing list) and OP-TEE should be also fix but in a relax way, allowing to register 0 sized shared memory buffers as previously supported. IMHO (that remains humble regarding compatibility vs fixup considerations).

Reply:
Shall I implement this relaxed way then?

I don't quite agree on the fact that is was "supported". It is not when the page offset is null. But it will indeed save older U-Boots from crashing upon such request.

What are the recommendations on backward compatibility between components that are part of the bootchain
?

Reply:
The fix posted to the U-Boot ML only addresses the OP-TEE services (devices) enumeration.
I did not fix the API function [`__tee_shm_add()`](https://source.denx.de/u-boot/u-boot/-/blob/v2023.10/drivers/tee/tee-uclass.c?ref_type=tags#L55) nor the optee `.shm_register` handler ([`optee_shm_register()`](https://source.denx.de/u-boot/u-boot/-/blob/v2023.10/drivers/tee/optee/core.c?ref_type=tags#L624)) that both currently obviously accept zero-sized buffers.
On Linux side, the generic `tee_shm_alloc_*()` functions call [`shm_alloc_helper()`](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/drivers/tee/tee_shm.c?h=v6.6#n77) that doesn't care about size value, not does the 2 optee driver handlers [`optee_shm_register()`](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/drivers/tee/optee/smc_abi.c?h=v6.6#n452) and [`optee_ffa_shm_register()`](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/drivers/tee/optee/ffa_abi.c?h=v6.6#n269). I belive they all delegate to other kernel routines to support or not the 0 page case.
On Linux userland client side, we trick the size forcing it to 8, which at some point can be seen as an abuse, could lead to registering 2 physical pages.
Do we need to address all those cases?